xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.grpc.Internal;
+import io.grpc.InternalLogId;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Serves as a wrapper for {@link XdsClientImpl} used on the server side by {@link
+ * io.grpc.xds.internal.sds.XdsServerBuilder}.
+ */
+@Internal
+public final class XdsClientWrapperForServerSds {
+  private static final Logger logger =
+      Logger.getLogger(XdsClientWrapperForServerSds.class.getName());
+
+  private static final EventLoopGroupResource eventLoopGroupResource =
+      Epoll.isAvailable() ? new EventLoopGroupResource("GrpcServerXdsClient") : null;
+  private final InternalLogId logId;
+  // Must be accessed from the syncContext
+  private boolean panicMode;
+  final SynchronizationContext syncContext =
```

Comment:
Not sure if creating a sync context here is appropriate. The whole control plane workflow should use a single sync context, it should not be something to be created just because you need it. You need to consider the whole server side flow.

Reply:
I looked at the client side and over there a sync context is [created](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L152) by a `ManagedChannelImpl`  and used everywhere starting from the name-resolver, load-balancers, and the XdsClient as well. So you have one sync context per managed channel. In this case (server side), there is currently no sync context being created because one wasn't needed on the server side so far. So I ended up creating it in this class which associates one sync context with one server (listener) and that cardinality seems correct. But may be the correct place to create this sync context is in the server builder and plumb the value all the way down here instead of creating it here. Let me know if you agree.

Reply:
Would need suggestion from @creamsoup

Reply:
@creamsoup lmk if you have comments here